blog/content/post/2024-04-12-odroid-m1-secure-boot.md

```diff
@@ -0,0 +1,598 @@
+
+#### Generating ramboot loader
+
+To check SecureMode state we need to run ramboot loader. To do that I used
+`boot_merger` tool with `RK3568MINIALL_RAMBOOT.ini` config file to create
+`rk356x_ramboot_loader_v1.21.108.bin`:
+
+```shell
+tools/boot_merger RKBOOT/RK3568MINIALL_RAMBOOT.ini
+Info:Pack loader ok.
+```
+
+#### Signing loader
+
+Now we need to sign generated `rk356x_ramboot_loader_v1.21.108.bin` loader.
+To do that I used `rk_sign_tool` from rkbin repository.
+First I needed to configure this tool with correct SoC and keys
+
+```shell
+ tools/rk_sign_tool cc --chip 3568 && tools/rk_sign_tool lk --key ../keys/dev.key --pubkey ../keys/dev.pubkey
+********sign_tool ver 1.4********
+set chip is 3568
+setting chip ok.
+********sign_tool ver 1.4********
+private key is ../keys/dev.key
+public key is ../keys/dev.pubkey
+loading key ok.
+```
+
+After which I signed loader.
+
+```shell
+tools/rk_sign_tool sl --loader rk356x_ramboot_loader_v1.21.108.bin
+********sign_tool ver 1.4********
+Loader is rk356x_ramboot_loader_v1.21.108.bin
+signing usbhead...
+failed to get key = sign_algo
+signing flashhead...
+failed to get key = sign_algo
+signing rk356x_ramboot_loader_v1.21.108 ok
+```
+
+Loader was signed correctly even though tool printed some failures.
+
+#### Sending ramboot loader
+
+To send ramboot loader I again used `upgrade_tool` but this time with `db`
+option.
+
+```shell
+./upgrade_tool db rkbin/rk356x_ramboot_loader_v1.21.108.bin
+```
+
+On UART console I had gotten this output
+
+```text
+Boot1 Release Time: Apr 20 2021 18:00:11, version: 1.08 USB BOOT
+ChipType = 0x18, 271
+SecureMode = 1
+atags_set_bootdev: ret:(0)
+UsbBoot ...684
+powerOn 931
+```
+
+Before writing hash to OTP memory ramboot loader returned `SecureMode = 0`.
+
+## Mainline U-Boot with signature verification
+
+Now that we have pre-loader (TPL + SPL) signature verification we can use
+pre-loader to verify next steps.
+
+To complete this stage I decided to use v2024.01
+[mainline U-Boot](https://github.com/3mdeb/u-boot/tree/2024.01-odroid-m1-sb-rk3568)
+with couple changes. I also needed rkbin repository and RSA keys and certificate
+(copied into u-boot directory). It's possible to use the same key as
+earlier or create new one.
+
+### U-Boot configuration
+
+The main changes needed in my commit were just adding `signature` and
+`u-boot-spl-pubkey-dtb` node in
+[`rockchip-u-boot.dtsi`](https://github.com/3mdeb/u-boot/blob/b5f0de18708112ce61a56526e6081796045e1763/arch/arm/dts/rockchip-u-boot.dtsi)
+and changing `CONFIG_SPL_STACK_R_MALLOC_SIMPLE_LEN` config variable to
+`0x150000` to fix `alloc space exhausted` error when booting.
+
+```text
+U-Boot SPL 2024.01-dirty (Jan 01 1970 - 00:00:00 +0000)
+Trying to boot from MMC2
+alloc space exhausted
+FIT buffer size: 1199104 bytes
+Could not get FIT buffer of 1199104 bytes
+        check CONFIG_SPL_SYS_MALLOC_SIZE
+```
+
+To configure U-Boot it's enough to use `odroid-m1-sb-rk3568_defconfig` and set
```

Comment:
Is this defconfig available publicly?

Reply:
051dcbef318bd22d15f61b72e7da8e782d1cd05d